### DISCUSSION TITLE
rebuilding-stongly-typed-id-as-a-source-generator-1-0-0-beta-release/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#111] [Announcements]
# rebuilding-stongly-typed-id-as-a-source-generator-1-0-0-beta-release/

In this post I describe the changes I made when rebuilding a library that auto-generates strongly-typed IDs for you to use source generators

https://andrewlock.net/rebuilding-stongly-typed-id-as-a-source-generator-1-0-0-beta-release/

**Comment by ItaiTzur76** [comment]
Hello, Andrew.
I started using this package and it is really wonderful.
I have a question, though - can a GUID-backed strongly-typed ID used as a primary-key in SQL-Server be auto-generated by the database?
I tried creating the following strongly-typed structure:
```cs
[SuppressMessage("Design", "CA1036:Override methods on comparable types", Justification = "Implementing IComparable is not necessary here")]
[StronglyTypedId(converters: StronglyTypedIdConverter.EfCoreValueConverter | StronglyTypedIdConverter.SystemTextJson | StronglyTypedIdConverter.TypeConverter)]
public partial struct DocumentId
{
}
```
Then I created the following entity:
```cs
[Index(nameof(Title), IsUnique = true)]
public sealed class Document
{
    public DocumentId Id { get; init; }

    public string Title { get; init; } = default!;
}
```
Then I added the following to the DbContext:
```cs
public DbSet<Document> Documents { get; init; } = default!;
```
Running the appropriate `dotnet ef migrations add` and then `dotnet ef database update` created the *documents* table in my SQL-Server database as expected. So I executed the following snippet:
```cs
Document document = new() { Title = "myDocument" };
DbContext.Documents.Add(document);
await DbContext.SaveChangesAsync();
```
This did create a row in that table but with `00000000-0000-0000-0000-000000000000` in the *id* field.
Is there a way to make SQL-Server auto-generate a random value for such a GUID-based strongly-typed field?

Regards,
Itai

[a comment by an automated account is not shown]
